### PR TITLE
POSIX compliance and new features

### PR DESCRIPTION
**Pull request by j8r**
I've added new improvements to your initial work:
- POSIX compliance accross shells
- Update tldr to its newest version
- Clear all cache files
- Download all cache files (for offline use)
- Improve readability with colors and tabs
- New one line command installation
- Use wget if curl is not present

I allowed myself to add myself on the license and the credits for my work. If you find this inappropriate, you can anyway remove my name but I will feel sad.

The one line command installation need to be tested when this PR will be merged.

### CODE REVIEW

**Comment by raylee** [comment]
Wow! You absolutely will get credit for your work, though I will have to read through it and test it first. Given my current schedule that may not happen until Saturday. I hope you don't mind the delay.

In the meantime, thank you for the PR!

**Comment by j8r** [comment]
No problem, thank you too for the fast reply! This will let me the time also to test more on my side since I've pushed to GitHub.

**Comment by j8r** [comment]
@raylee Do you plan do review this soon?

**Comment by tpo** [comment]
@raylee - I have read through j8r's commits in his fork [1] until commit 5edba24 [2]) and I think his commits are OK.

In general I hate doing `wget http://some.url/some.shell.script.sh | sh` which j8r added in his `README.md/documentation`, but I think that's up to your taste.

Also I very much prefer your way of installing `tldr` to `~/bin`, which is also how I organize my scripts. Even though I am also using aliases sometimes I think aliases are evil (since 1. they are not in your `$PATH` and so `vim $(which some_alias)` will not work among other things and 2. they can not be called from other scripts).

Other than that I think j8r's patches are fine and I appreciate his effort to make `tldr` run on other POSIX shells out there - I imagine there are BSD people that will appreciate that. 

I did not an "in depth" review neither of your code nor of j8r's patches, just quickly read through everything from start to end.

So here you have my 2 cents. Best greets and thanks to both of you @raylee and @j8r,
*t

[1] https://github.com/j8r/tldr
[2] https://github.com/j8r/tldr/commit/5edba24cf78f6095ce28aaa86e23777627253c96

**Comment by j8r** [comment]
@tpo Thanks for your review. :)

For the autoinstall script, I understand that it can be a security risk, but it is very practical for new comers. Therefore I've also added manual instructions. Overall everything is up to the end user.

I quite disagree with you for `~/bin` because tldr is not only an executable but also a directory with subfolders with a cache. For me a `~./tldr` is a good directory, with inside `tldr.sh` and the `cache |` folder.

Overall, I haven't encountered major issues but I've observed that when the `cache` folder is not present and we run tldr (that will create the cache), nothing is printed. We need to run it again.

File: README.md [hunk 1]
```diff
@@ -1,35 +1,55 @@
...
 
-![tldr screenshot](Screenshot.png?raw=true)
+![tldr screenshot](screenshot.png?raw=true)
 
 # Setup
 
-	mkdir -p ~/bin
-	curl -o ~/bin/tldr https://raw.githubusercontent.com/raylee/tldr/master/tldr
-	chmod +x ~/bin/tldr
+`wget -qO- https://raw.githubusercontent.com/DFabric/DPlatform-ShellCore/master/install.sh | sh`
```
Comment (raylee) [inline README.md]: I'm uncomfortable with this even being an option. I don't want to encourage this sort of installation method.

File: README.md [hunk 2]
```diff
@@ -1,35 +1,55 @@
...
 
-	command
-		Show examples for this command
+		$ tldr tar
+		$ tldr -p linux tar
+```
 
 The client caches a copy of all pages and the index locally under
-~/.config/tldr. By default, the cached copies will expire in 14 days.
+~/.tldr. By default, the cached copies will expire in 14 days.
```
Comment (raylee) [inline README.md]: I'd rather keep this .config/tldr. This is the freedesktop.org standard for where config directories should be stored. If there were a .rc file that we wanted to store (for user configuration), that would be appropriate in ~/ (ie ~/.tldr as a file, not a directory).

File: install.sh [hunk 3]
```diff
@@ -0,0 +1,27 @@
+#!/bin/sh
+
+
+[ ! -d ~/.tldr ] && printf "tldr seems to have already been previously installed.\nDelete the \'$HOME/.tldr\' directory first before relaunching this script\n" && exit 1 || mkdir -p ~/.tldr/cache
```
Comment (raylee) [inline install.sh]: One of the goals I tried to maintain with the script is that it should "just work" at all times. So the script should clean up after itself and maintain its own caches. This is another benefit of using ~/.config/tldr/, as it is more clearly a directory under the script's control.

File: install.sh [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+download() {
```
Comment (raylee) [inline install.sh]: I like the idea of a download command that covers all the cases, nice work!

File: install.sh [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  hash curl 2>/dev/null && curl -sf -o "$1" "$2" || { hash wget 2>/dev/null && wget -qO "$1" "$2"; } || echo "Fail to download $2" && exit 1
```
Comment (raylee) [inline install.sh]: But I'm confused as to why you have `hash` in front of the commands. I don't think it's necessary, is it?

File: install.sh [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+# Download tldr.sh and index.json
+download ~/.tldr/tldr.sh https://raw.githubusercontent.com/raylee/tldr/master/tldr.sh
+download ~/.tldr/cache/index.json http://tldr-pages.github.io/assets/index.json
+
+# Add alias to rc files
+add_alias() {
```
Comment (raylee) [inline install.sh]: I'm uncomfortable with this. Another goal for this script was such that it could be included in distros directly with no more dependencies than the core files. Part of that means being a good citizen with the standards they ask, which includes not automatically modifying the user's profile files. So I think this belongs either in the manual instructions or as an 'install' target on on the command line which does all the magic.

As an aside, these aliases are unnecessary in the original instructions, as tldr is a command directly in their path:

`curl -o ~/bin/tldr https://raw.githubusercontent.com/raylee/tldr/master/tldr`

File: tldr.sh [hunk 7]
```diff
@@ -1,58 +1,56 @@
-#!/bin/bash
+#!/bin/sh
 # tldr client by Ray Lee, http://github.com/raylee/tldr
+# Improvements added by Julien Reichardt https://github.com/j8r
 # a Sunday afternoon's project, I'm sure there's room for improvement. PRs welcome!
-
-set -uf -o pipefail
```
Comment (raylee) [inline tldr.sh:5]: Why are you removing this?

File: tldr.sh [hunk 8]
```diff
@@ -1,58 +1,56 @@
...
-    [ -d ~/.config ] && configdir=~/.config/tldr
+    configdir=~/.tldr/cache
 
     platform=$(get_platform)
     base_url="https://raw.githubusercontent.com/tldr-pages/tldr/master/pages"
     index_url="http://tldr-pages.github.io/assets/index.json"
     index="$configdir/index.json"
     cache_days=14
-    force_update=''
+    force_update=
```
Comment (raylee) [inline tldr.sh]: Did you test this change?

File: tldr.sh [hunk 9]
```diff
@@ -72,57 +72,60 @@ init_term_cmds() {
...
-        italic=$(echo -e "\033[3m")
-        eitalic=$(echo -e "\033[23m")
+        italic=$(printf "\033[3m")
+        eitalic=$(printf "\033[23m")
     }
 }
 
 heading() {
     local line="$*"
-    echo "$bold$red${line:2}$reset"
```
Comment (raylee) [inline tldr.sh:89]: So, when I originally wrote the coloring and formatting I was matching the official tldr output. Maybe I got it wrong? Same comment with the other color and style changes below.

File: tldr.sh [hunk 10]
```diff
@@ -72,57 +72,60 @@ init_term_cmds() {
...
 list_item() {
     local line="$*"
-    echo "$line$reset"
+    printf "  $green$line$reset"
 }
 
 code() {
     local line="$*"
     # I'm sure there's a better way to strip the first and last characters.
-    line="${line:1}"
```
Comment (raylee) [inline tldr.sh:105]: Can you provide some commentary to explain why you're changing this? It works correctly already, so not sure why you're changing this?

File: tldr.sh [hunk 11]
```diff
@@ -72,57 +72,60 @@ init_term_cmds() {
...
+    line="${line#?}"
     line="${line%\`}"
     # convert {{variable}} to italics
-    line=${line//\{\{/$italic}
-    line=${line//\}\}/$eitalic}
+    # Check if the shell support buil-in substitutions
+    (${a//}) 2>/dev/null && line=${line//\{\{/$reset$white} || { hash sed 2>/dev/null && line="$(printf "$line" | sed "s/{{/$reset$white/g" )"; }
```
Comment (raylee) [inline tldr.sh]: Nicely done. Same comment about `hash` though.

File: tldr.sh [hunk 12]
```diff
@@ -72,57 +72,60 @@ init_term_cmds() {
...
 }
 
 # an idiot-level recognition of tldr's markdown. Needs improvement, or
 # subcontracting out to a markdown -> ANSI formatting command
 display_tldr() {
     # read one line at a time, don't strip whitespace ('IFS='), and process
     # last line even if it doesn't have a newline at the end
-    while IFS= read -r line || [[ -n "$line" ]]; do
-        start=${line:0:1}           # get the first character
+    while read line || [ -n "$line" ]; do
```
Comment (raylee) [inline tldr.sh]: So this change remove the IFS= which prevents while from stripping whitespace. Rationale?

File: tldr.sh [hunk 13]
```diff
@@ -156,49 +159,62 @@ path_for_cmd() {
...
 get_tldr() {
     local p="$(path_for_cmd $1)"
     cached="$configdir/$p"
-    recent "$cached" || {
-        mkdir -p $(dirname $cached)
-        curl -sf -o "$cached" "$base_url/$p"
-    }
-    # if the curl failed for some reason, keep cat from whinging
-    cat "$cached" 2>/dev/null
+    recent "$cached" || { mkdir -p $(dirname $cached); download "$cached" "$base_url/$p"; }
```
Comment (raylee) [inline tldr.sh]: Please split across multiple lines as per the original.

File: tldr.sh [hunk 14]
```diff
@@ -207,21 +223,32 @@ exit 0
...
+        -h|-\?|--help)
+            usage
+            ;;
+        -d|--download)
+            pindex=$(cat $configdir/index.json)
+            parse
+            ;;
         -l|--list)
-            >&2 echo -e "Known tldr pages: \n"
-            tr '{' '\n' < "$configdir/index.json" | cut -d '"' -f4 | column
```
Comment (raylee) [inline tldr.sh:212]: Fair enough, I guess the user can pipe it to `column` on their own.

File: tldr.sh [hunk 15]
```diff
@@ -233,14 +260,18 @@ do
...
 
 if [ -z "$tldr" ]; then
-    echo "tldr page for command $page not found"
+    printf "tldr page for command $page not found
+Try updating with \"tldr --update\", or submit a pull request to:
+https://github.com/tldr-pages/tldr\n"
     exit 1
 fi
 
-display_tldr <<< "$tldr"
```
Comment (raylee) [inline tldr.sh:245]: For my own education, is <<< not supported in all POSIX shells?

File: install.sh [hunk 16]
```diff
@@ -0,0 +1,30 @@
+#!/bin/sh
```
Comment (raylee) [inline install.sh]: Another unstated goal was that everything would be a single file, which keeps the installation easy. I could see adding all the alias stuff to an install target in tldr itself, (tldr -i).

**Review by raylee — CHANGES_REQUESTED** [review]
Lots of really good work! I like a lot of the directions you are heading, but I do have a bunch of detailed comments on the PR.

I think many of them are simply due to my lack of shell scripting experience. So please keep that in mind when you read my comments, I think I'll just need a bit of education as to why you made some of the changes as they aren't immediately apparent to me.

Thank you for giving such care and love to this little weekender project!

File: README.md (hunk 2 above)
Comment (tpo) [reply README.md]: I was actually reflecting about this as well. Not only is ~/.config a freedesktop.org standard (which I appreciate), but there's also ~/.cache from the same standard.

I was thinking about whether the cached tldr *.md files should possibly go to ~/.cache/tldr. I think they actually should. I consider stuff in ~/.cache "ad hoc reproducible" stuff - here in the sense that it can be re-fetched at any time, if internet is available.

Also ~/.cache is something that I would not want to backup, which is the case here for *.md pages that can be trivially and painlessly re-downloaded from the tldr git repo.

So I'd argue that the cached *.md pages should actually go into ~/.cache/tldr instead.

File: tldr.sh (hunk 14 above)
Comment (tpo) [reply tldr.sh:212]: even after 20 years of unix `column` is a command that I'm still only half aware of.

File: README.md (hunk 2 above)
Comment (j8r) [reply README.md]: I agree we can do this it was my first idea, but I've finally chosen this way to have a true portable script, that we can bring on any directory and execute it, with a `cache` folder that will be created.

File: install.sh (hunk 3 above)
Comment (j8r) [reply install.sh]: Indeed it's not very needed, I've deleted this in a further commit (commit [bb3a00658f74f49c3271621e467193b91c2a9241](https://github.com/j8r/tldr/commit/bb3a00658f74f49c3271621e467193b91c2a9241))

File: README.md (hunk 2 above)
Comment (raylee) [reply README.md]: ~/.cache/tldr/ makes more sense, let's go with that.

File: install.sh (hunk 5 above)
Comment (j8r) [reply install.sh]: I think it's better to check if `curl` or `wget` are present to avoid the user to install the one or the other package. Often `wget` is already present by default, but if `curl` is present, better to use it because it's faster (for me at least) and I also prefer it, but this is only my personal opinion.

File: tldr.sh (hunk 14 above)
Comment (raylee) [reply tldr.sh:212]: @tpo Heh, I find it one of those 'obscure but occasionally useful' commands, along with `join` and a few others.

File: tldr.sh (hunk 7 above)
Comment (j8r) [reply tldr.sh:5]: Because this cause some shells like `dash` (`set: Illegal option -o pipefail`) . It appears to be not POSIX compliant, and I haven't noticed problems after deleting, so I remove it.

File: tldr.sh (hunk 8 above)
Comment (j8r) [reply tldr.sh]: Yes, this reset the same way the variable (from 'yes').

File: tldr.sh (hunk 10 above)
Comment (j8r) [reply tldr.sh:105]: Because of `dash` again. I'm not sure if this substitution is POSIX or not (or what verison) — `bash --posix` and `ksh` works but a `shellcheck` shows that it isn't POSIX.
Anyway, I remove the first and the latest "`" character. We can replace "${line%\'`'}" by "${line%?}" (just tested).

File: tldr.sh (hunk 11 above)
Comment (j8r) [reply tldr.sh]: Thanks, but i'm not specially proud of this hack. I've found that `(${a//})` return an error on `dash`, but not on other modern shells. Here we can remove the `hash sed 2>/dev/null` check because this utility is virtually present in all UN*X machine (even more than `wget` that is a "standalone" package)

File: tldr.sh (hunk 12 above)
Comment (j8r) [reply tldr.sh]: Now I'm also asking myself this questions. It's not really needed... It's nevertheless a good practice?

File: tldr.sh (hunk 15 above)
Comment (j8r) [reply tldr.sh:245]: You are right :)

File: README.md (hunk 1 above)
Comment (j8r) [reply README.md]: I really would like tldr to be the most accessible to the wider range of people.
I agree to not encourage this practice, but please keep (not highlighted if you want) this feature for people (I admit like me) who just wants to have the things done as efficiently as possible.
This is not as evil (at least here), specially for such an open source script that isn't ran privileged.

File: tldr.sh (hunk 9 above)
Comment (j8r) [reply tldr.sh:89]: You have interpreted the markdown strictly "as is". I've followed the [actual official tldr image with its colors](https://raw.githubusercontent.com/tldr-pages/tldr/master/screenshot.png), that interpret this character changes to colors, that I think is a more human readable way to transmit informations.
I've taken the liberty to make some little changes, specially at the top for the command title and its description. I think we can better differentiate them from the main contents.
Overall this colors changes are opinionated and the final thought is up to each one.

File: README.md (hunk 2 above)
Comment (j8r) [reply README.md]: Ok good. I've an idea, we can add a `--portable|-p` argument with no cache, or in `/tmp`, it will solves the problem!

File: install.sh (hunk 6 above)
Comment (j8r) [reply install.sh]: Agree, specially because `tldr` is a single standalone file. By the way I thinks it's would be better on `/usr/bin` when tldr will have it's package, but actually even more on `/usr/local/bin`.

File: README.md (hunk 2 above)
Comment (j8r) [reply README.md]: For me we should change the argument `-p|--platform`  to `-o|--os` like on other tldr implementations like [here](https://github.com/tldr-pages/tldr-node-client/blob/master/bin/tldr#L22). We could also add `--linux`,  `--osx` and `--sunos`.

File: tldr.sh (hunk 14 above)
Comment (tpo) [reply tldr.sh:212]:
```
$ tldr join
tldr page for command join not found
```

damn!

**Comment by j8r** [comment]
I've made various changes according to your requests, thank you @raylee to have build this project from scratch, on which I have based my work. Thanks to you and @tpo  to take time to review my code!

I've kept `install.sh` and `README.md`unmodified, modify them as your convenience.
I've also added a `-p|--portable` argument for those who want to execute `tldr` on the go.

**Comment by j8r** [comment]
@raylee and @tpo, I've finalized the changes you have requested. Tell me if the PR is good to merge now, thank you again!

**Comment by raylee** [comment]
Good morning! I'll be reviewing these changes in the next couple days, thank you for both your patience and your continued efforts.

**Comment by j8r** [comment]
@raylee Have you reviewed the changes?

**Comment by raylee** [comment]
I've given it a first pass and it looks good. The last thing I'd like to think about is making the upgrade clean for users, in particular removing the old cached pages if they exist in the previous location. Thoughts?

**Comment by j8r** [comment]
Also if we use `upgrade` when`tldr` is in `/usr/local/bin`, we need to be root or change the owner of the file.
By the way I agree, we can do this like you said, and also add a `remove` feature if you are OK.

**Comment by raylee** [comment]
You've been patient enough on this, let's leave the upgrade issues for a separate PR. So for now, I have no requests for changes.

**Comment by raylee** [comment]
So I just want to check with you on your intent. Here's a screenshot of the rendering before the changes in this PR:

image

And here it is afterward:

image

(This is with the same source md file in each case, naturally.) Does this match your intent? Some of the color choices are difficult with a bright background.

There is another open PR that I'm going to look at which makes the color choices configurable via environment variables. If it's okay with you, I'm going to cherry-pick the changes from your fork, and allow you to override the colors via the other PR.

**Comment by j8r** [comment]
We should create a bright/dark terminal mode, because we will have problems like this, good point.
Also I've closely matched the "*official*" colors (including bright ones).
Anyway I think it's simpler to merge this, because the modifications brought by this PR are quite numerous compared to the color patch.
Then we can work with @ewels to solve this color issue, and *cherry-pick* his fixes.
But the final decision is yours.

**Comment by j8r** [comment]
I'm just thinking, most macOS users (like you) have bright terminal and most of Linux users (like me) have dark ones.
So why not use this to set right defaults to the colors, of course in addition of the @ewels 's PR?

**Comment by raylee** [comment]
Cherry-picking isn't a big deal. There's no need for a themes, just stick with all color choices of the same contrast/saturation. My editor/terminal color schemes change by time of day (and system I'm ssh'd into), and the color choices in the original version, while dull, work just fine with both backgrounds.

Anyway, there's a reason why it's usually one change per PR. Let's aim for that in the future. Thanks!

**Comment by j8r** [comment]
Ok sure!
There are pros and cons of both configurations.
Nice output but not universal, or universal dull output?
IMHO we should open an issue on this to let the users say what they really would like.
This would be interesting to see how other tldr clients handle this.

**Comment by ewels** [comment]
I think boring but customisable colours is going to be the best way to go. If someone tries this script and can't read the output on the first go, they're probably not going to hang around trying to figure out how to change them. However, making it possible (and preferably simple) to change the colours is definitely a worthwhile feature.

If the readme has a bunch of "themes" with screenshots and cut+paste env variables, then that should be fine I think? Or even named presets such as `dark` and `light` etc. Could save this with the cache instead of in environment variables? Sorry, just thinking out loud here :)

**Comment by tpo** [comment]
@ewels - let's keep in mind, that tldr is also and defacto "just a little shell script" where opening it in a text editor, scrolling through the script, finding the place where the colors are set and changing them right there does not amount to rocket science, espacially if the place where the color setting happens would be nicely commented...

**Comment by ewels** [comment]
Sorry @tpo, but I disagree. For me the entire spirit of _tldr_ is that it's beginner friendly. Editing a bash script that's hundreds of lines long so that you can read the output doesn't really seem to fit that ethos. Having a handful of cut and paste `export VAR` statements in the readme isn't really adding much complexity to the code and keeps things simple for those who are new to a command line environment.
